presto-native-execution/presto_cpp/main/PrestoServerOperations.cpp

```diff
@@ -242,11 +243,56 @@ std::string PrestoServerOperations::taskOperation(
 
 std::string PrestoServerOperations::serverOperation(
     const ServerOperation& op,
-    proxygen::HTTPMessage* /* unused */) {
+    proxygen::HTTPMessage* message) {
   switch (op.action) {
-    case ServerOperation::Action::kTrace: {
+    case ServerOperation::Action::kTrace:
       return velox::process::TraceContext::statusLine();
-    }
+    case ServerOperation::Action::kSetState:
+      if (server_) {
+        const auto& stateStr = message->getQueryParam("state");
+        const auto prevState = server_->nodeState();
+        NodeState newNodeState{NodeState::kActive};
+        if (stateStr == "active") {
+          newNodeState = NodeState::kActive;
+        } else if (stateStr == "inactive") {
+          newNodeState = NodeState::kInActive;
+        } else if (stateStr == "shutting_down") {
+          newNodeState = NodeState::kShuttingDown;
+        } else {
+          return fmt::format(
+              "Invalid state '{}'. "
+              "Supported states are: 'active', 'inactive', 'shutting_down'. "
+              "Example: server/setState?state=shutting_down",
+              stateStr);
+        }
+        if (newNodeState != prevState) {
+          LOG(INFO) << "Setting node state to "
+                    << nodeState2String(newNodeState);
+          server_->setNodeState(newNodeState);
+        }
+        return fmt::format(
+            "New node state: '{}', previous state: '{}'.",
+            nodeState2String(newNodeState),
+            nodeState2String(prevState));
+      }
+      return "No PrestoServer to change state of (it is nullptr).";
+    case ServerOperation::Action::kAnnouncer:
+      if (server_) {
+        const auto& actionStr = message->getQueryParam("action");
+        if (actionStr == "enable") {
+          server_->enableAnnouncer(true);
+          return "Announcer enabled";
+        } else if (actionStr == "disable") {
+          server_->enableAnnouncer(false);
+          return "Announcer disabled";
+        }
+        return fmt::format(
```

Comment:
Can we report errors using HTTP 500 ?

Reply:
I don't see why we need this. It is not error, as such.

Reply:
@mbasmanova 
I'm changing the PR according to some comment.
I was looking to address yours, but it will require a bit more intrusive change to the whole PrestoServerOperations class.

Should we do this in the follow up?

Reply:
@spershin Sergey, thank you for considering my comments. A follow-up would be fine with me.